### PR TITLE
Create Tag API

### PR DESCRIPTION
**Pull request by SoSorryTT**
## PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Others. (Please describe)

### Changes

1. Create Tag API     
2. Create GET DELETE UPDATE CREATE for Tag API

### CODE REVIEW

File: garnbarn_api/urls.py [hunk 1]
```diff
@@ -2,10 +2,13 @@
 from django.db.models import base
 from django.urls.conf import include
 from rest_framework.routers import DefaultRouter
-from .views import AssignmentViewset
+
```
[the remark this answers is not in the document]
Comment (SoSorryTT) [reply garnbarn_api/urls.py]: Done in commit a0572a0

File: garnbarn_api/views.py [hunk 2]
```diff
@@ -1,6 +1,8 @@
+from functools import partial
```
[the remark this answers is not in the document]
Comment (SoSorryTT) [reply garnbarn_api/views.py]: It's was  a mistake so i remove it in commit a0572a0

File: garnbarn_api/serializer.py [hunk 3]
```diff
@@ -94,3 +94,33 @@ class Meta:
...
+
+class UpdateTagApiSerializer(serializers.ModelSerializer):
+    """Serializer for the Update Tag API"""
+    class Meta:
+        model = Tag
+        fields = ['id',
+                'name',
+                'color'
+                ]
+        depth = 1
```
[the remark this answers is not in the document]
Comment (SoSorryTT) [reply garnbarn_api/serializer.py]: The create one use for create tag and check by serializer. The update one use for update tag and check by serializer.

File: garnbarn_api/models.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 from django.db.models.deletion import SET_NULL
 from django.utils import timezone
 import math
-import uuid
+from django.db.models.deletion import CASCADE
```
Comment (sirateek) [inline garnbarn_api/models.py:7]: I think this line is unnecessary

File: garnbarn_api/tests/test_tag_model.py [hunk 5]
```diff
@@ -1,13 +1,39 @@
...
+        """Test that that has no color."""
+        self.assertEqual(None, self.tag.color)
+
+    def test_tag_with_no_reminder(self):
+        """Test tag that has no reminder time."""
+        self.assertEqual(None, self.tag.reminder_time)
+
+    def test_tag_reminder(self):
+        """Test tag remider."""
+        self.assertEqual([3600, 1800], self.tag2.reminder_time)
```
Comment (sirateek) [inline garnbarn_api/tests/test_tag_model.py:39]: The reminder time must be sorted before storing it in the database. I think this test case is wrong.

**Review by sirateek — CHANGES_REQUESTED** [review]
There is still no test for Tag API

**Comment by sirateek** [comment]
@SoSorryTT Please also update the Document of the API in the gh-pages branch

File: garnbarn_api/models.py (hunk 4 above)
Comment (SoSorryTT) [reply garnbarn_api/models.py:6]: Remove it in commit 852d87d.

File: garnbarn_api/tests/test_tag_model.py (hunk 5 above)
Comment (SoSorryTT) [reply garnbarn_api/tests/test_tag_model.py:39]: Just to test the result of the reminderTime the data will be sort in serializer(Done in commit 852d87d.).

File: garnbarn_api/tests/test_tag_serializer.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+from django.test import TestCase
+from garnbarn_api.models import Tag
+from garnbarn_api.serializer import CreateTagApiSerializer
+
+
+class TagAPITest(TestCase):
```
Comment (sirateek) [inline garnbarn_api/tests/test_tag_serializer.py:6]: Update the Class name, This is the test for Serializer

File: garnbarn_api/tests/test_view.py [hunk 7]
```diff
@@ -217,6 +226,101 @@ def test_delete(self):
...
+        response = self.client.get("/api/v1/tag/2/")
+        converted_data = convert_to_json(response.content)
+        expected_result = json.dumps({
+            "id": 2,
+            "name": "test_tag2",
+            "author": {
+                'line': None,
+                'name': 'user_name',
+                'uid': 'user_id'
+                },
```
Comment (sirateek) [inline garnbarn_api/tests/test_view.py:240]: Why author returns the entire data for the user?

File: garnbarn_api/tests/test_view.py [hunk 8]
```diff
@@ -217,6 +226,101 @@ def test_delete(self):
...
+        """Create tag object with no error."""
+        data = {
+            'name': "test_tag",
+            'author': self.user,
+            'color': '#4285F4',
+            "reminderTime": [3600, 1800],
+            'subscriber': self.user
+        }
+        response = self.client.post("/api/v1/tag/", data)
+        self.assertEqual(200, response.status_code)
```
Comment (sirateek) [inline garnbarn_api/tests/test_view.py:284]: The reminderTime should be sorted before storing in the database

**Review by sirateek — CHANGES_REQUESTED** [review]

**Comment by sirateek** [comment]
Let ignore the Travis check for PR since there is a conflict in the migration file.

Follow these steps.
1. Merge this branch into iteration2
2. Remove all migration files in `terartion2` branch.
3. Commit the new migration files into `iteration2` branch

To avoid duplication works.

File: garnbarn_api/models.py [hunk 9]
```diff
@@ -1,10 +1,9 @@
-from typing import AbstractSet, Optional
 from django.db import models
 import datetime
 from django.db.models.deletion import CASCADE, SET_NULL
 from django.utils import timezone
 import math
-from django.db.models.deletion import CASCADE
+import uuid
```
[the remark this answers is not in the document]
Comment (SoSorryTT) [reply garnbarn_api/models.py]: I remove it already but it keep coming back remove done in d34d7a6.

File: garnbarn_api/tests/test_tag_serializer.py (hunk 6 above)
Comment (SoSorryTT) [reply garnbarn_api/tests/test_tag_serializer.py]: Change done in commit d34d7a6.

File: garnbarn_api/tests/test_view.py (hunk 7 above)
Comment (SoSorryTT) [reply garnbarn_api/tests/test_view.py]: Remove done in commit d34d7a6.

File: garnbarn_api/tests/test_view.py (hunk 8 above)
Comment (SoSorryTT) [reply garnbarn_api/tests/test_view.py]: Done in commit d34d7a6.

**Review by abounk — APPROVED** [review]
LGTM 👍😊

**Pull request merged**